### PR TITLE
TN-1772 - Adding Timothy install to CLI

### PR DESCRIPTION
**Pull request by daniel-tonic**

### CODE REVIEW

File: cli/cmd/timstall/timstall.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+func doUninstall(
+	reader io.Reader,
+	writer io.Writer,
+	cloneSettings *tt.CloneSettings,
+	chartValues *tt.TimChartValues,
+	installValues *tt.TonicInstallValues,
+) (error) {
+	ctx := context.Background()
```
Comment (anr-tonic) [inline cli/cmd/timstall/timstall.go:216]: We do this in several spots. What do you think adding something like the following to helpers?

```go
package helpers

import (
	"context"
	"os"
	"os/signal"
)

func StopWithSigInt(f func(context.Context) error, ctx context.Context) error {
	thisCtx, cancel := context.WithCancel(ctx)
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)

	defer func() {
		signal.Stop(c)
		cancel()
	}()

	go func() {
		select {
		case <-c:
			cancel()
		case <-ctx.Done():
		}
	}()

	return f(thisCtx)
}
```

And used like:

```go
err = h.StopWithSigInt(
    func(ctx context.Context) error { 
        return doUninstall(reader, writer, cloneSettings, chartValues, installValues, ctx) 
    },
    context.Background(),
)
```

File: cli/pkg/chart/timinstall/database_options.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    if values.Password == "" {
+        metadataDbPassword, err := passwordGenerator.GeneratePassword(32)
+        if err != nil {
+            metadataDbPassword = "toniciscool"
+        }
+        values.Password = metadataDbPassword
+    }
+
+    values.Database = "tim"
+    values.Host = fmt.Sprintf("%s-postgres", TimothyPgReleaseName)
```
Comment (anr-tonic) [inline cli/pkg/chart/timinstall/database_options.go:257]: This needs to be `postgresql`, same issue that I patched the other day.

File: cli/pkg/kubernetes/helpers.go [hunk 3]
```diff
@@ -20,3 +20,47 @@ func GetPodsForNamespace(kubeConfig *KubeConfig, ctx context.Context, namespace
 	pods, podsErr := client.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
 	return pods, podsErr
 }
+
+func GetKubernetesContexts(cloneSettings *tt.CloneSettings) ([]string, *KubeConfig, error) {
+	kubeConfigPath := filepath.Join(cloneSettings.UserHomeDir, ".kube", "config")
```
Comment (anr-tonic) [inline cli/pkg/kubernetes/helpers.go:25]: Seeing this reminded me that I wanted to run TIM w/ a different kubernetes config and it wasn't possible. I've filed a ticket for supporting using the `KUBECONFIG` envvar if it is set over this path but marked it as low priority.

**Review by anr-tonic — COMMENTED** [review]
This is just an end of the day cursory review, I'll dig in more depth tomorrow.

File: cli/pkg/helpers/display.go [hunk 4]
```diff
@@ -11,3 +11,6 @@ var YELLOW = color.New(color.FgYellow).SprintFunc()
 var BOLD = color.New(color.Bold).SprintFunc()
 
 var TONIC = BOLD(CYAN("TOИIC"))
+var TIM = BOLD(CYAN("TIM"))
```
[the remark this answers is not in the document]
Comment (daniel-tonic) [reply cli/pkg/helpers/display.go:14]: I actually flipped all the letters.

File: cli/pkg/kubernetes/collect_install_values.go [hunk 5]
```diff
@@ -36,10 +37,10 @@ func CollectClusterInstallValues(
 		})
 
 		collector.CollectStringValue(c.StringValueOptions[tt.TonicInstallValues]{
-			Prompt: fmt.Sprintf("What namespace should %s be installed in?", h.TONIC),
+			Prompt: fmt.Sprintf("What namespace should %s be installed in?", productDisplay),
 			GetCurrentValue: func(v *tt.TonicInstallValues) *string {
 				if v.KubernetesNamespace == "" {
-					return h.Ptr("tonic")
+					return h.Ptr(strings.ToLower(string(productName)))
```
[the remark this answers is not in the document]
Comment (daniel-tonic) [reply cli/pkg/kubernetes/collect_install_values.go:43]: The compiler doesn't seem to like that. :(

File: cli/pkg/chart/installers/postgres.go [hunk 6]
```diff
@@ -105,6 +106,55 @@ func (installer *postgresInstaller) install(installValues tt.TonicInstallValues,
...
+	_, err = client.Run(installer.releaseName)
+	return err
+}
+
+func (installer *postgresInstaller) upgrade(installValues tt.TonicInstallValues, chartValues tt.CommonChartValues, ctx context.Context) error {
+	if chartValues.GetInClusterDbValues() == nil {
+		return nil
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (daniel-tonic) [reply cli/pkg/chart/installers/postgres.go:145]: No idea.  The function used to be
```
func (installer *postgresInstaller) upgrade(installValues tt.TonicInstallValues, chartValues tt.TonicChartValues, ctx context.Context) error {
	if chartValues.InClusterDbValues == nil {
		return nil
	}
	return nil
}
```

I didn't even think about it.  I can get rid of the conditional.

File: cli/pkg/chart/installers/allos.go [hunk 7]
```diff
@@ -78,7 +87,7 @@ func (installer *allosInstaller) Install(installValues tt.TonicInstallValues, ch
 	}
 }
 
-func (installer *allosInstaller) install(installValues tt.TonicInstallValues, chartValues tt.TonicChartValues, ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (daniel-tonic) [reply cli/pkg/chart/installers/allos.go]: done

File: cli/pkg/chart/installers/postgres.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (daniel-tonic) [reply cli/pkg/chart/installers/postgres.go]: updated

File: cli/cmd/timstall/timstall.go (hunk 1 above)
Comment (daniel-tonic) [reply cli/cmd/timstall/timstall.go]: I like it. Done and done.

File: cli/pkg/chart/timinstall/database_options.go (hunk 2 above)
Comment (daniel-tonic) [reply cli/pkg/chart/timinstall/database_options.go]: I looked for all the places where this string is being assembled and replaced it with a helper method so we shouldn't have problems with this staying in sync anymore.

File: cli/pkg/chart/installers/postgres.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anr-tonic) [reply cli/pkg/chart/installers/postgres.go]: lol wtf was I thinking then.

**Review by anr-tonic — APPROVED** [review]

**Pull request merged**